File: orion-cost-svc-go/internal/service/cost_service.go
package service

import (
	"context"
	"errors"
	"time"

	"orion/cost-svc-go/internal/models"
	"orion/cost-svc-go/internal/repository"

	"go.uber.org/zap"
)

var (
	ErrInvalidTenantID = errors.New("tenant ID required")
	ErrNotFound        = errors.New("record not found")
)

// CostService handles cost aggregation and querying by resource, time, and tenant.
type CostService struct {
	repo   *repository.CostRepository
	logger *zap.Logger
}

// NewCostService creates a new cost service instance.
func NewCostService(repo *repository.CostRepository, logger *zap.Logger) *CostService {
	return &CostService{repo: repo, logger: logger}
}

// RecordCost inserts a cost record for a tenant.
func (s *CostService) RecordCost(ctx context.Context, req *models.RecordCostRequest) error {
	if req.TenantID == "" {
		return ErrInvalidTenantID
	}

	date := time.Now()
	if req.Date != "" {
		parsed, err := time.Parse("2006-01-02", req.Date)
		if err == nil {
			date = parsed
		}
	}

	resourceID := (*string)(nil)
	if req.ResourceID != "" {
		resourceID = &req.ResourceID
	}

	region := (*string)(nil)
	if req.Region != "" {
		region = &req.Region
	}

	record := &models.CostRecord{
		TenantID:   req.TenantID,
		Date:       date,
		Service:    req.Service,
		ResourceID: resourceID,
		Region:     region,
		Cost:       req.Cost,
		Currency:   req.Currency,
		Category:   req.Category,
		Tags:       req.Tags,
	}

	err := s.repo.CreateCostRecord(ctx, record)
	if err != nil {
		s.logger.Error("failed to create cost record", zap.Error(err))
		return err
	}

	s.logger.Info("cost record created", zap.String("tenant_id", req.TenantID), zap.Float64("cost", req.Cost))
	return nil
}

// ListCosts retrieves paginated cost records with optional filters.
func (s *CostService) ListCosts(ctx context.Context, tenantID string, filter *models.ListCostsRequest, offset, limit int) ([]models.CostRecord, error) {
	repoFilter := &repository.ListFilter{
		StartDate:  filter.StartDate,
		EndDate:    filter.EndDate,
		Service:    filter.Service,
		ResourceID: filter.ResourceID,
		Region:     filter.Region,
	}

	return s.repo.FindCostRecords(ctx, tenantID, repoFilter, offset, limit)
}

// GetTotalCost returns the total cost for a tenant in a date range.
func (s *CostService) GetTotalCost(ctx context.Context, tenantID, startDate, endDate string) (float64, error) {
	return s.repo.GetTotalCost(ctx, tenantID, startDate, endDate)
}

// GetCostByService returns cost aggregated by service.
func (s *CostService) GetCostByService(ctx context.Context, tenantID, startDate, endDate string) ([]models.CostAggregation, error) {
	return s.repo.GetCostByService(ctx, tenantID, startDate, endDate)
}

// GetCostByResource returns cost aggregated by resource.
func (s *CostService) GetCostByResource(ctx context.Context, tenantID, startDate, endDate string) ([]models.CostAggregation, error) {
	return s.repo.GetCostByResource(ctx, tenantID, startDate, endDate)
}

// GetCostSummary returns aggregated cost summary across dimensions.
func (s *CostService) GetCostSummary(ctx context.Context, tenantID, startDate, endDate string) (*models.CostSummary, error) {
	return s.repo.GetCostSummary(ctx, tenantID, startDate, endDate)
}

// GetCostTrend returns daily cost trend for a period.
func (s *CostService) GetCostTrend(ctx context.Context, tenantID, startDate, endDate string) (*models.CostTrendResult, error) {
	points, err := s.repo.GetDailyCostTrend(ctx, tenantID, startDate, endDate)
	if err != nil {
		return nil, err
	}

	if len(points) == 0 {
		return &models.CostTrendResult{
			Points: points,
			TotalCost: 0,
			AverageCost: 0,
			Trend: "stable",
			ChangeRate: 0,
		}, nil
	}

	totalCost := 0.0
	for _, p := range points {
		totalCost += p.Cost
	}
	averageCost := totalCost / float64(len(points))

	trend := calculateTrend(points)
	changeRate := calculateChangeRate(points)

	return &models.CostTrendResult{
		Points:      points,
		TotalCost:   roundFloat(totalCost),
		AverageCost: roundFloat(averageCost),
		Trend:       trend,
		ChangeRate:  roundFloat(changeRate),
	}, nil
}

func calculateTrend(points []models.CostTrendPoint) string {
	if len(points) < 2 {
		return "stable"
	}

	mid := len(points) / 2
	firstHalf := points[:mid]
	secondHalf := points[mid:]

	firstAvg := sumOf(firstHalf) / float64(len(firstHalf))
	secondAvg := sumOf(secondHalf) / float64(len(secondHalf))

	if firstAvg > 0 {
		rate := (secondAvg - firstAvg) / firstAvg * 100
		if rate > 10 {
			return "increasing"
		} else if rate < -10 {
			return "decreasing"
		}
	}
	return "stable"
}

func calculateChangeRate(points []models.CostTrendPoint) float64 {
	if len(points) < 2 {
		return 0
	}
	mid := len(points) / 2
	firstHalf := points[:mid]
	secondHalf := points[mid:]
	firstAvg := sumOf(firstHalf) / float64(len(firstHalf))
	secondAvg := sumOf(secondHalf) / float64(len(secondHalf))
	if firstAvg > 0 {
		return (secondAvg - firstAvg) / firstAvg * 100
	}
	return 0
}

func sumOf(points []models.CostTrendPoint) float64 {
	sum := 0.0
	for _, p := range points {
		sum += p.Cost
	}
	return sum
}

// CountCostRecords returns the total number of cost records for a tenant.
func (s *CostService) CountCostRecords(ctx context.Context, tenantID string) (int64, error) {
	return s.repo.CountCostRecords(ctx, tenantID)
}

File: legacy/orion-platform-service-ts/src/services/cost/CostService.ts
/**
 * CostService - Business logic layer for Cost/FinOps operations
 */

import { CostRepository, CostRecord, Budget, CostAggregation } from './CostRepository';

export interface ListCostOptions {
  page?: number;
  limit?: number;
  tenantId?: string;
  startDate?: Date;
  endDate?: Date;
  service?: string;
}

export interface PaginatedResult<T> {
  data: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export class CostServiceError extends Error {
  constructor(message: string, public code: string) { super(message); this.name = 'CostServiceError'; }
}

export class CostService {
  private repository: CostRepository;
  constructor(repository: CostRepository) { this.repository = repository; }

  async recordCost(tenantId: string, service: string, cost: number, options?: { date?: Date; resourceId?: string; region?: string; tags?: Record<string, string> }): Promise<CostRecord> {
    if (!tenantId) throw new CostServiceError('Tenant ID required', 'INVALID_INPUT');
    return this.repository.createCostRecord(tenantId, options?.date || new Date(), service, cost, options?.resourceId, options?.region, options?.tags);
  }

  async getCosts(options: ListCostOptions = {}): Promise<PaginatedResult<CostRecord>> {
    const { page = 1, limit = 20, tenantId, startDate, endDate, service } = options;
    const offset = (page - 1) * limit;
    const [costs, _] = await Promise.all([
      this.repository.findAll({ tenantId, startDate, endDate, service, limit, offset }),
      Promise.resolve(0)
    ]);
    return { data: costs, total: costs.length, page, limit, totalPages: Math.ceil(costs.length / limit) };
  }

  async getCostByService(tenantId: string, startDate: Date, endDate: Date): Promise<CostAggregation[]> {
    return this.repository.getCostByService(tenantId, startDate, endDate);
  }

  async getTotalCost(tenantId: string, startDate: Date, endDate: Date): Promise<number> {
    return this.repository.getTotalCost(tenantId, startDate, endDate);
  }

  async createBudget(tenantId: string, name: string, amount: number, period: string, alertThreshold: number): Promise<Budget> {
    if (!tenantId) throw new CostServiceError('Tenant ID required', 'INVALID_INPUT');
    return this.repository.createBudget(tenantId, name, amount, period, alertThreshold);
  }

  async listBudgets(tenantId?: string): Promise<Budget[]> {
    return this.repository.findAllBudgets(tenantId);
  }

  async getBudget(id: string): Promise<Budget> {
    const budget = await this.repository.findBudgetById(id);
    if (!budget) throw new CostServiceError(`Budget not found: ${id}`, 'NOT_FOUND');
    return budget;
  }

  async getBudgetAlerts(tenantId: string): Promise<Budget[]> {
    return this.repository.getBudgetAlerts(tenantId);
  }

  async updateBudgetSpend(id: string): Promise<Budget> {
    const budget = await this.repository.findBudgetById(id);
    if (!budget) throw new CostServiceError(`Budget not found: ${id}`, 'NOT_FOUND');
    
    const now = new Date();
    const startDate = new Date(now.getFullYear(), now.getMonth(), 1);
    const total = await this.repository.getTotalCost(budget.tenant_id, startDate, now);
    
    const updated = await this.repository.updateBudgetSpend(id, total);
    return updated!;
  }
}